src/Kernels/MetropolisHastingsLazy.hs

```diff
@@ -4,69 +4,67 @@ import Control.Monad.State.Lazy
-acceptanceProbability current proposal logProb =
-  min 1 (exp (logProb proposal - logProb current))
+acceptanceProbability :: Double -> Double -> Double
+acceptanceProbability current proposal =
+  min 1 (exp (Helpers.logProb proposal - Helpers.logProb current))
 
-propose :: (Num a, UniformRange a) => a -> a -> State StdGen a
-propose x stepsize = state $ do
-  (proposal, newGen) <- uniformR ((x - stepsize), (x + stepsize))
+propose :: Double -> State StdGen Double
+propose x = state $ do
+  (proposal, newGen) <- uniformR (x - stepsize, x + stepsize)
   return (proposal, newGen)
- 
-acceptOrReject :: (Num a, Num b, UniformRange b, Floating b, Ord b) => a -> a -> (a -> b) -> State StdGen a 
-acceptOrReject current proposal logProb = state $ do
+  where
+    stepsize = 0.9
+
+propose' :: State (StdGen, Double) Double
+propose' = do
+  (gen, current) <- get
+  let (proposal, newGen) = uniformR (current - stepsize, current + stepsize) gen
+  put(newGen, current)
+  return proposal
+  where
+    stepsize = 0.9
+
+acceptOrReject :: Double -> Double -> State StdGen Double
+acceptOrReject current proposal = state $ do
   (rand, newGen) <- uniformR (0.0, 1.0)
-  if rand < acceptanceProbability current proposal logProb 
+  if rand < acceptanceProbability current proposal
    then return (proposal, newGen)
    else return (current, newGen)
 
-transition :: (Num a, Num b, Ord b, UniformRange a, UniformRange b, Floating b) => a -> (a -> b) -> a -> State StdGen a
-transition current logProb stepsize = 
-  propose' >>= acceptOrReject'
-  where
-    propose' = propose current stepsize
-    acceptOrReject' = \proposal -> acceptOrReject current proposal logProb
-
-data MhData = MhData {
-  n :: Int,
-  chain :: [Double],
-  gen :: StdGen,
-  logProb :: Double -> Double,
-  stepsize :: Double
-}
+acceptOrReject' :: Double -> State (StdGen, Double) Double
+acceptOrReject' proposal = do
+  (gen, current) <- get
+  let (rand, newGen) = uniformR (0.0, 1.0) gen
+  let nextValue = if rand < acceptanceProbability current proposal
+                   then proposal
+                   else current
+  put (newGen, nextValue)
+  return nextValue
 
-simpleSample :: MhData -> MhData
-simpleSample dat
-  | length (chain dat) == n dat = dat
-  | otherwise = simpleSample newDat
+transition :: Double -> State StdGen Double
+transition current =
+  propose' >>= acceptOrReject'
   where
-    stateVal = transition (head $ chain dat) (logProb dat) (stepsize dat)
-    (newVal, newGen) = runState stateVal (gen dat)
-    newDat = dat {gen = newGen, chain = newVal : (chain dat)}
-
-d = MhData {n = 5, chain = [1.0], gen = (mkStdGen 42), logProb = Helpers.logProb, stepsize = 0.1}
-res = simpleSample d
+    propose' = propose current
+    acceptOrReject' = acceptOrReject current
 
+-- State s a
+-- State (StdGen, Double) Double 
+transition' :: State (StdGen, Double) Double
+transition' = propose' >>= acceptOrReject'
```

Comment:
After having changed `propose'` and `acceptOrReject'`, we need to partially apply them here to make the composition work.

Reply:
done